src/Sylius/Bundle/ResourceBundle/Routing/ResourceLoader.php

```diff
@@ -0,0 +1,169 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Bundle\ResourceBundle\Routing;
+
+use Gedmo\Sluggable\Util\Urlizer;
+use Sylius\Component\Resource\Metadata\MetadataInterface;
+use Sylius\Component\Resource\Metadata\RegistryInterface;
+use Symfony\Component\Config\Loader\LoaderInterface;
+use Symfony\Component\Config\Loader\LoaderResolverInterface;
+use Symfony\Component\Yaml\Yaml;
+
+/**
+ * @author Paweł Jędrzejewski <pawel@sylius.org>
+ */
+class ResourceLoader implements LoaderInterface
+{
+    /**
+     * @var RegistryInterface
+     */
+    private $resourceRegistry;
+
+    /**
+     * @var RouteFactoryInterface
+     */
+    private $routeFactory;
+
+    /**
+     * @param RegistryInterface $resourceRegistry
+     * @param RouteFactoryInterface $routeFactory
+     */
+    public function __construct(RegistryInterface $resourceRegistry, RouteFactoryInterface $routeFactory)
+    {
+        $this->resourceRegistry = $resourceRegistry;
+        $this->routeFactory = $routeFactory;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function load($resource, $type = null)
+    {
+        $configuration = Yaml::parse($resource);
+
+        if (isset($configuration['only']) && isset($configuration['except'])) {
+            throw new \InvalidArgumentException('You can configure only one of "except" & "only" options.');
+        }
+
+        $routesToGenerate = array('show', 'index', 'create', 'update', 'delete');
+
+        if (isset($configuration['only'])) {
+            $routesToGenerate = $configuration['only'];
+        }
+        if (isset($configuration['except'])) {
+            $routesToGenerate = array_diff($routesToGenerate, $configuration['except']);
+        }
+
+        $isApi = $type === 'sylius.resource_api';
+
+        $metadata = $this->resourceRegistry->get($configuration['alias']);
+        $routes = $this->routeFactory->createRouteCollection();
+
+        $rootPath = sprintf('/%s/', isset($configuration['path']) ? $configuration['path'] : Urlizer::urlize($metadata->getPluralName()));
+
+        if (in_array('index', $routesToGenerate)) {
+            $indexRoute = $this->createRoute($metadata, $configuration, $rootPath, 'index', array('GET'));
+            $routes->add($this->getRouteName($metadata, $configuration, 'index'), $indexRoute);
```

Comment:
`generateRouteIfDefined` or `generateRouteIfNeeded`

```php

function generateRouteIfDefined($metadata, $configuration, $actionName, $rootPath, $methods)
{
    if (in_array($actionName, $routesToGenerate)) {
        $route = $this->createRoute($metadata, $configuration, $rootPath, $actionName, $methods);
        $routes->add($this->getRouteName($metadata, $configuration, $actionName), $route);
    }
}
```

so in load you will have:

```php
$this->generateRouteIfDefined($metadata, $configuration, 'index', $rootPath, array('GET'));
$this->generateRouteIfDefined($metadata, $configuration, 'show', $rootPath . '{id}', array('GET'));
$this->generateRouteIfDefined($metadata, $configuration, 'delete', $rootPath . '{id}', array('DELETE'));
$this->generateRouteIfDefined($metadata, $configuration, 'create', $isApi ? $rootPath : $rootPath . 'new', $isApi ? array('POST') : array('GET', 'POST'));
...
```

Much more readable IMO. @pjedrzejewski WDYT?